File: cane-cpp/cane-core/src/runtime_load_plan.cpp
#include "cane/runtime_load_plan.hpp"
#include "runtime_data_internal.hpp"
#include "operation.hpp"

namespace cane::detail {
struct RuntimeLoadPlanState {
    format::Json document;
    RuntimeLoadOptions options;
    std::vector<RuntimeTextureRequest> requests;
    std::vector<RuntimeLoadWarning> warnings;
    std::string load_operation;
};
}
namespace cane {
namespace {
std::shared_ptr<const detail::RuntimeLoadPlanState> prepare(format::Json document, const RuntimeLoadOptions& options,
    std::vector<RuntimeLoadWarning> warnings, const char* load_operation) {
    using format::ModelValidation;
    auto state = std::make_shared<detail::RuntimeLoadPlanState>();
    state->document = std::move(document); state->options = options; state->warnings = std::move(warnings); state->load_operation = load_operation;
    const auto& source = state->document;
    const auto index = ModelValidation(source, load_operation).validate();
    for (const auto& supplied : options.atlas_json) if (index.catalogs[0].find(supplied.first) == index.catalogs[0].end())
        throw Error(ErrorCode::invalid_argument, load_operation, "Supplied Atlas ID is not declared by this Runtime document.", "atlasId", supplied.first);
    for (const auto& row : source["images"]) if (ModelValidation::get(row, "atlasId").is_null()) {
        RuntimeTextureRequest request; request.image_id = row["imageId"].get<std::string>(); request.path = row["path"].get<std::string>();
        if (row.contains("width")) request.declared_width = row["width"].get<std::uint32_t>();
        if (row.contains("height")) request.declared_height = row["height"].get<std::uint32_t>();
        state->requests.push_back(std::move(request));
    }
    for (const auto& reference : source["atlases"]) {
        const auto id = reference["atlasId"].get<std::string>(); const auto found = options.atlas_json.find(id);
        if (found == options.atlas_json.end()) throw Error(ErrorCode::missing_resource, load_operation, "Declared Atlas JSON is required.", "atlasId", id);
        const auto atlas = AtlasData::from_json(found->second);
        if (atlas.atlas_id() != id) throw Error(ErrorCode::validation_failed, load_operation, "Atlas document identity differs from its reference.", "atlasId", id);
        for (const auto& page : atlas.pages()) {
            RuntimeTextureRequest request; request.kind = TextureResourceKind::atlas_page;
            request.atlas_id = id; request.page_id = page.page_id; request.atlas_path = reference["path"].get<std::string>(); request.path = page.image_path;
            request.declared_width = page.width; request.declared_height = page.height; request.color_space = atlas.color_space(); request.alpha_mode = atlas.alpha_mode();
            request.min_filter = page.min_filter; request.mag_filter = page.mag_filter; request.wrap_u = page.wrap_u; request.wrap_v = page.wrap_v;
            state->requests.push_back(std::move(request));
        }
    }
    return state;
}
}
RuntimeLoadPlan RuntimeLoadPlan::from_json(std::string_view utf8, const RuntimeLoadOptions& options) {
    return detail::operation("prepareJson", [&] { return RuntimeLoadPlan(prepare(format::parse_json(utf8), options, {}, "loadJson")); });
}
std::vector<RuntimeAtlasReference> RuntimeLoadPlan::inspect_atlas_references(const std::vector<std::uint8_t>& source) {
    return detail::operation("inspectDependencies", [&] {
        const bool binary = source.size() >= 5 && std::string(source.begin(), source.begin() + 5) == "CANEB";
        auto document = binary ? format::decode_caneb(source).document : format::parse_json(std::string(source.begin(), source.end()));
        (void)format::ModelValidation(document, "inspectDependencies").validate();
        std::vector<RuntimeAtlasReference> references;
        for (const auto& row : document["atlases"])
            references.push_back({row["atlasId"].get<std::string>(), row["path"].get<std::string>()});
        return references;
    });
}
RuntimeLoadPlan RuntimeLoadPlan::from_caneb(const std::vector<std::uint8_t>& bytes, const RuntimeLoadOptions& options) {
    return detail::operation("prepareCaneb", [&] {
        auto decoded = format::decode_caneb(bytes); std::vector<RuntimeLoadWarning> warnings;
        for (const auto& warning : decoded.warnings) warnings.push_back({warning.code, warning.operation, warning.section_tag});
        return RuntimeLoadPlan(prepare(std::move(decoded.document), options, std::move(warnings), "loadCaneb"));
    });
}
const std::vector<RuntimeTextureRequest>& RuntimeLoadPlan::texture_requests() const noexcept { return state_->requests; }
const std::vector<RuntimeLoadWarning>& RuntimeLoadPlan::warnings() const noexcept { return state_->warnings; }
RuntimeData RuntimeLoadPlan::create_data(const std::vector<DecodedTextureDimensions>& decoded) const {
    return detail::operation("completeLoad", [&] {
        auto options = state_->options; options.decoded_textures = decoded;
        auto data = detail::RuntimeDataAccess::load(state_->document, options, state_->warnings, state_->load_operation.c_str());
        if (state_->options.decoded_textures) data.validate_decoded_texture_sizes(*state_->options.decoded_textures);
        return data;
    });
}
}

File: cane-csharp/cane-core/Runtime/RuntimeLoadPlan.cs
using System;
using System.Collections.Generic;
using Cane.Format;

namespace Cane
{
    /// <summary>Owned source dependency metadata; file acquisition remains in the host.</summary>
    public sealed class RuntimeAtlasReference
    {
        public string AtlasId { get; }
        public string Path { get; }
        internal RuntimeAtlasReference(string atlasId, string path) { AtlasId = atlasId; Path = path; }
    }

    /// <summary>A renderer-neutral decode request. Missing declared dimensions stay null until the host observes pixels.</summary>
    public sealed class RuntimeTextureRequest
    {
        public TextureResourceKind Kind { get; }
        public string? ImageId { get; }
        public string? AtlasId { get; }
        public string? PageId { get; }
        public string? AtlasPath { get; }
        public string Path { get; }
        public int? DeclaredWidth { get; }
        public int? DeclaredHeight { get; }
        public string ColorSpace { get; } = "srgb";
        public string AlphaMode { get; } = "straight";
        public string PixelFormat { get; } = "rgba8";
        public string MinFilter { get; } = "linear";
        public string MagFilter { get; } = "linear";
        public string WrapU { get; } = "clamp";
        public string WrapV { get; } = "clamp";
        internal RuntimeTextureRequest(Json image)
        {
            Kind = TextureResourceKind.Direct; ImageId = image.S("imageId"); Path = image.S("path");
            DeclaredWidth = image["width"].IsNull ? (int?)null : image.I("width");
            DeclaredHeight = image["height"].IsNull ? (int?)null : image.I("height");
        }
        internal RuntimeTextureRequest(Json reference, Json atlas, Json page)
        {
            Kind = TextureResourceKind.AtlasPage; AtlasId = atlas.S("atlasId"); PageId = page.S("pageId");
            AtlasPath = reference.S("path"); Path = page.S("image"); DeclaredWidth = page.I("width"); DeclaredHeight = page.I("height");
            ColorSpace = atlas.S("colorSpace"); AlphaMode = atlas.S("alphaMode"); PixelFormat = page.S("pixelFormat");
            MinFilter = page.S("minFilter"); MagFilter = page.S("magFilter"); WrapU = page.S("wrapU"); WrapV = page.S("wrapV");
        }
    }

    /// <summary>Owned preparation input, not a playable RuntimeData. Complete model/resource validation precedes CreateData success.</summary>
    public sealed class RuntimeLoadPlan
    {
        private readonly Json document;
        private readonly Dictionary<string, string> atlases;
        private readonly string[] optionalSections;
        private readonly DecodedTextureDimensions[]? expectedDimensions;
        private readonly bool compatibilityOption;
        public IReadOnlyList<RuntimeTextureRequest> TextureRequests { get; }
        public IReadOnlyList<RuntimeWarning> Warnings { get; }

        /// <summary>Inspect declared atlas paths before supplying atlas documents to the load plan.
        /// This validates the source container/header and references; CreateData still performs full validation.</summary>
        public static IReadOnlyList<RuntimeAtlasReference> InspectAtlasReferences(byte[] source) => Wrap("inspectDependencies", () => {
            if (source == null) throw new ArgumentNullException(nameof(source));
            bool binary = source.Length >= 5 && source[0] == 'C' && source[1] == 'A' && source[2] == 'N' && source[3] == 'E' && source[4] == 'B';
            Json doc = binary ? Caneb.Decode(source, out _) : Json.Parse(source);
            Validation.Root(doc);
            var references = new List<RuntimeAtlasReference>();
            var ids = new HashSet<string>(StringComparer.Ordinal);
            foreach (Json row in doc["atlases"].Items) {
                Validation.Closed(row, "atlasId path", "atlasId path");
                string id = Validation.Id(row, "atlasId"), path = row.S("path");
                Validation.Path(path);
                if (!ids.Add(id)) throw new RuntimeException(RuntimeErrorCode.ValidationFailed, "inspectDependencies", "Duplicate atlas ID.", "atlasId", id);
                references.Add(new RuntimeAtlasReference(id, path));
            }
            return (IReadOnlyList<RuntimeAtlasReference>)references.AsReadOnly();
        });

        public static RuntimeLoadPlan FromJson(string json, RuntimeLoadOptions? options = null) =>
            Wrap("prepareJson", () => new RuntimeLoadPlan(Json.Parse(json), options ?? new RuntimeLoadOptions(), Array.Empty<string>()));
        public static RuntimeLoadPlan FromJson(byte[] utf8, RuntimeLoadOptions? options = null) =>
            Wrap("prepareJson", () => new RuntimeLoadPlan(Json.Parse(utf8), options ?? new RuntimeLoadOptions(), Array.Empty<string>()));
        public static RuntimeLoadPlan FromCaneb(byte[] bytes, RuntimeLoadOptions? options = null) =>
            Wrap("prepareCaneb", () => { Json doc = Caneb.Decode(bytes, out string[] sections); return new RuntimeLoadPlan(doc, options ?? new RuntimeLoadOptions(), sections); });

        private RuntimeLoadPlan(Json doc, RuntimeLoadOptions options, string[] sections)
        {
            document = doc; Validation.Root(doc);
            if (options.AtlasJson == null) throw new RuntimeException(RuntimeErrorCode.InvalidArgument, "prepare", "Atlas inputs are required.", "atlasJson");
            atlases = new Dictionary<string, string>(options.AtlasJson, StringComparer.Ordinal);
            compatibilityOption = options.AllowUnverifiedFeatures;
            expectedDimensions = options.DecodedTextures == null ? null : RuntimeData.CopyDecodedTextureFacts(options.DecodedTextures, "prepare");
            optionalSections = (string[])sections.Clone(); var warnings = new RuntimeWarning[sections.Length];
            for (int i = 0; i < sections.Length; ++i) warnings[i] = new RuntimeWarning(sections[i]); Warnings = Array.AsReadOnly(warnings);
            var requests = new List<RuntimeTextureRequest>();
            foreach (Json image in doc["images"].Items) {
                Validation.Image(image);
                if (image["atlasId"].IsNull) requests.Add(new RuntimeTextureRequest(image));
            }
            var referenced = new HashSet<string>(StringComparer.Ordinal);
            foreach (Json reference in doc["atlases"].Items) {
                Validation.Closed(reference, "atlasId path", "atlasId path"); Validation.Path(reference.S("path"));
                string id = Validation.Id(reference, "atlasId"); referenced.Add(id);
                if (!atlases.TryGetValue(id, out string? source)) throw new RuntimeException(RuntimeErrorCode.MissingResource, "prepare", "Atlas JSON is required.", "atlasId", id);
                Json atlas = Json.Parse(source); Validation.Atlas(atlas);
                if (atlas.S("atlasId") != id) throw new RuntimeException(RuntimeErrorCode.ValidationFailed, "prepare", "Atlas document and reference identities differ.", "atlasId", id);
                foreach (Json page in atlas["pages"].Items) requests.Add(new RuntimeTextureRequest(reference, atlas, page));
            }
            foreach (string id in atlases.Keys) if (!referenced.Contains(id))
                throw new RuntimeException(RuntimeErrorCode.InvalidArgument, "prepare", "Supplied Atlas ID is not declared.", "atlasId", id);
            TextureRequests = requests.AsReadOnly();
        }

        public RuntimeData CreateData(IReadOnlyList<DecodedTextureDimensions> decoded) => Wrap("completeLoad", () => {
            var observations = RuntimeData.CopyDecodedTextureFacts(decoded, "completeLoad");
            RuntimeData data = RuntimeData.Load(document, new RuntimeLoadOptions { AtlasJson = atlases, DecodedTextures = observations, AllowUnverifiedFeatures = compatibilityOption });
            if (expectedDimensions != null) data.ValidateDecodedTextureCatalog(expectedDimensions);
            data.SetLoadWarnings(optionalSections); return data;
        });
        private static T Wrap<T>(string operation, Func<T> body)
        {
            try { return body(); }
            catch (RuntimeException e) { throw new RuntimeException(e.Code, operation, e.Message, e.Field, e.EntityId); }
            catch (OutOfMemoryException) { throw new RuntimeException(RuntimeErrorCode.ResourceLimit, operation, "Unable to allocate load preparation data."); }
            catch (Exception e) when (e is ArgumentException || e is OverflowException || e is IndexOutOfRangeException)
            { throw new RuntimeException(RuntimeErrorCode.ValidationFailed, operation, e.Message); }
        }
    }
}
